### PR TITLE
Params for Observables + ... Wire Notation

### PR DESCRIPTION
**Pull request by puzzleshark**
This PR does the following

* Allows Observables to contain parameters within observable definitions.
* Implements the ... wire notation for declarations.
* Updates the Transformer to use `v_args(inline=True)` decorator (this combined with using `[expr]` instead of `expr?` in .lark file and `maybe_placeholders=True` option in the parser settings allows for much simpler transformer methods!
* Creates a debug and non-debug mode to run parser in, non debug mode doesn't build parse tree, and increases parse performance in some cases by 30%!
* Refactors `terms` to `factors`

### CODE REVIEW

**Comment by trbromley** [comment]
[sc-10100]

File: xir/program.py [hunk 1]
```diff
@@ -137,27 +137,55 @@ def use_floats(self) -> bool:
...
+        self.params = list(params or [])
+        self.wires = tuple(wires or [])
+
+    def __str__(self) -> str:
+        wires = ", ".join([str(v) for v in self.wires])
+        if len(self.params) == 0:
+            return f"{self.name}[{wires}]"
+        else:
+            params = ", ".join([str(v) for v in self.params])
+            return f"{self.name}({params})[{wires}]"
```
Comment (Mandrenkov) [inline xir/program.py:160]: The logic here is fine although I have a few code style suggestions:
1. A clean usage of `map` should be preferred to list comprehensions. For example, compare
    ```python
    wires = ", ".join([str(v) for v in self.wires])
    ```
    to
    ```python
    wires = ", ".join(map(str, self.wires))
    ```
2. It is more Pythonic to check if a sequence is empty using
    ```python
    if self.params:
    ```
3. There is no need to use `else` when the final line of an `if` block is `return`.

Putting it all together,

```suggestion
        wires = ", ".join(map(str, self.wires))
        if self.params:
            return f"{self.name}[{wires}]"

        params = ", ".join(map(str, self.params))
        return f"{self.name}({params})[{wires}]"
```

File: xir/program.py [hunk 2]
```diff
@@ -137,27 +137,55 @@ def use_floats(self) -> bool:
...
+
+
 class ObservableStmt:
     """Observable statements to be used in observable definitions.
 
     Args:
         pref (Decimal, int, str): prefactor to the observable terms
-        terms (list): list of observables and the wire(s) they are applied to
+        factors (list[ObservableFactor]): list of observable factors
```
Comment (Mandrenkov) [inline xir/program.py:168]: Does this need to be a list? Or can we get away with making this a [`Sequence`](https://docs.python.org/3/library/collections.abc.html#collections-abstract-base-classes) or something instead?

File: xir/program.py [hunk 3]
```diff
@@ -137,27 +137,55 @@ def use_floats(self) -> bool:
...
-        self._terms = terms
+        self._factors = factors
 
         self._use_floats = use_floats
 
     def __str__(self) -> str:
         """Serialized string representation of an ObservableStmt."""
-        terms = [f"{t[0]}[{t[1]}]" for t in self.terms]
-        terms_as_string = " @ ".join(terms)
+        factors_as_string = " @ ".join([str(factor) for factor in self._factors])
```
Comment (Mandrenkov) [inline xir/program.py:185]: See comment above.
```suggestion
        factors_as_string = " @ ".join(map(str, self._factors))
```

File: xir/program.py [hunk 4]
```diff
@@ -167,9 +195,9 @@ def pref(self) -> Union[Decimal, float, int, str]:
         return self._pref
 
     @property
-    def terms(self) -> List:
+    def factors(self) -> List:
```
Comment (Mandrenkov) [inline xir/program.py:198]: Assuming that `self._factors` is always a list,
```suggestion
    def factors(self) -> List[ObservableStmt]:
```
Otherwise, if it is a sequence, we can write
```suggestion
    def factors(self) -> Sequence[ObservableStmt]:
```

File: xir/program.py [hunk 5]
```diff
@@ -507,7 +535,13 @@ def add_observable(
...
                 "Replacing old definition with new definition."
             )
-
+        if wires is None:
+            wires = []
+            for statement in statements:
+                for factor in statement.factors:
+                    for wire in factor.wires:
+                        if wire not in wires:
+                            wires.append(wire)
```
Comment (Mandrenkov) [inline xir/program.py]: Are you sure this is needed? I thought we accounted for this in `Transformer.obs_def`.

File: xir/__init__.py [hunk 6]
```diff
@@ -1,13 +1,40 @@
...
+    "ObservableFactor",
+    "ObservableStmt",
+    "Program",
+    "Statement",
+    "Transformer",
+]
 
 
-def parse_script(script: str, **kwargs) -> Program:
+def parse_script(script: str, debug=True, **kwargs) -> Program:
```
Comment (Mandrenkov) [inline xir/__init__.py:20]: We should add a type hint for the debug flag and describe its effect in the docstring.
```suggestion
def parse_script(script: str, debug: bool =True, **kwargs) -> Program:
```

File: tests/test_program.py [hunk 7]
```diff
@@ -5,6 +5,7 @@
 from typing import Any, Dict, Iterable, List, MutableSet, Sequence
 
 import pytest
+from xir import ObservableFactor
```
Comment (Mandrenkov) [inline tests/test_program.py:8]: Our style convention is to write `xir.ObservableFactor` instead of just `ObservableFactor`.
```suggestion
```

File: xir/parser.py [hunk 8]
```diff
@@ -174,70 +173,28 @@ def expr(self, args):
     # definitions and statements
     ##############################
 
-    def gate_def(self, args):
+    @v_args(inline=True)
```
Comment (Mandrenkov) [inline xir/parser.py:177]: TIL!

File: xir/parser.py [hunk 9]
```diff
@@ -174,70 +173,28 @@ def expr(self, args):
...
-            elif isinstance(arg, Statement):
-                if has_declared_wires:
-                    stmts = args[i:]
-                    break
-                stmts.append(arg)
-
-                int_wires = [w for w in arg.wires if isinstance(w, int)]
-                if int_wires and max(int_wires) > max_wire:
-                    max_wire = max(int_wires)
+        if wires is not None and len(wires) > 0:
```
Comment (Mandrenkov) [inline xir/parser.py]: I think this suffices:
```suggestion
        if wires:
```

File: xir/parser.py [hunk 10]
```diff
@@ -174,70 +173,28 @@ def expr(self, args):
...
-                wires = tuple(range(max_wire + 1))
-            else:
-                wires = tuple(dict.fromkeys(wires))
+            wires = tuple(range(max_wire + 1))
+        else:
+            # remove duplicate wires while maintaining order
+            wires = tuple(dict.fromkeys(wires))
 
-        self._program.add_observable(name, params, wires, stmts)
+        self._program.add_gate(name, [] if params_list is None else params_list[1], wires, stmts)
```
Comment (Mandrenkov) [inline xir/parser.py:193]: Generally, it's cleaner to leave `x if y else z` out of long expressions:
```suggestion
        params =  [] if params_list is None else params_list[1]
        self._program.add_gate(name, params, wires, stmts)
```

File: xir/parser.py [hunk 11]
```diff
@@ -275,69 +232,74 @@ def application_stmt(self, args):
...
-        pref = simplify_math(args[0])
-        terms = args[1]
-        return ObservableStmt(pref, terms, use_floats=self.use_floats)
+        if wires is None:
+            max_wire = 0
+            for stmt in statements:
+                for factor in stmt.factors:
+                    int_wires = [w for w in factor.wires if isinstance(w, int)]
+                    if int_wires and max(int_wires) > max_wire:
+                        max_wire = max(int_wires)
```
Comment (Mandrenkov) [inline xir/parser.py:240]: The `if` statement is redundant here:
```suggestion
                    max_wire = max(max_wire, *int_wires)
```

File: xir/parser.py [hunk 12]
```diff
@@ -275,69 +232,74 @@ def application_stmt(self, args):
...
-        return [(args[i], args[i + 1]) for i in range(0, len(args) - 1, 2)]
+        params = tuple(params[1]) if params is not None else tuple()
+        self._program.add_observable(name, params, wires, statements)
+
+    def obs_stmt_list(self, stmts):
+        return stmts
+
+    @v_args(inline=True)
+    def obs_stmt(self, coeff, factors):
+        return ObservableStmt(simplify_math(coeff), factors, use_floats=self.use_floats)
```
Comment (Mandrenkov) [inline xir/parser.py:255]: It would be nice to stick with one nomenclature for the pre-factor/coefficient of an observable statement.
```suggestion
    def obs_stmt(self, pref, factors):
        return ObservableStmt(simplify_math(pref), factors, use_floats=self.use_floats)
```

File: xir/parser.py [hunk 13]
```diff
@@ -4,22 +4,21 @@
...
 
 import lark
+from lark import v_args
 
 from .decimal_complex import DecimalComplex
-from .program import Declaration, ObservableStmt, Program, Statement
+from .program import Declaration, ObservableStmt, Program, Statement, ObservableFactor
 from .utils import simplify_math
 
+ENUM_WIRES = "ENUM_WIRES"
```
Comment (Mandrenkov) [inline xir/parser.py:13]: I don't think we really need this constant.
```suggestion
```

File: xir/parser.py [hunk 14]
```diff
@@ -275,69 +232,74 @@ def application_stmt(self, args):
...
-        if len(args) == 2:
-            name, params = args[0], args[1][1]
-        else:
-            name = args[0]
-            params = []
+    def wire_list(self, args):
+        return args
+
+    def ENUM(self, _):
+        return ENUM_WIRES
```
Comment (Mandrenkov) [inline xir/parser.py:286]: Why not return the actual `...` object itself?
```suggestion
        return ...
```

**Review by Mandrenkov — CHANGES_REQUESTED** [review]
Awesome work, Sam! 🎉

One thing I would like to see before approving this PR is more tests. Specifically, I'd like to have confidence that the XIR parser can correctly handle `...` (and any edge cases) as well as observable factors that are applied to more than one wire.

File: xir/parser.py (hunk 11 above)
Comment (puzzleshark) [reply xir/parser.py:240]: hmm not sure that it is. This is actually copied code from the gate case

File: xir/parser.py (hunk 14 above)
Comment (puzzleshark) [reply xir/parser.py:286]: not sure if it's a good idea to have identical string literals in multiple places

File: xir/xir.lark [hunk 15]
```diff
@@ -13,23 +13,29 @@ option: name ":" val
 // declarations
 ?declaration: (gate_decl | obs_decl | func_decl | out_decl)
 
-gate_decl: _GATE name params_list? wires ";"
-obs_decl: _OP name params_list? wires ";"
-out_decl: _OUT name params_list? wires ";"
-func_decl: _FUNC name params_list? ";"
+gate_decl: _GATE name [params_list] wires_or_enum ";"
```
[the remark this answers is not in the document]
Comment (puzzleshark) [reply xir/xir.lark:16]: `[params_list]` will return `None` if it is not found. Works will in combination with `v_args(inline=True)`

File: xir/xir.lark [hunk 16]
```diff
@@ -77,6 +83,7 @@ _OP: "obs"
 _OUT: "out"
 INVERSE: "inv"
 CTRL: "ctrl"
+ENUM: "..."
```
[the remark this answers is not in the document]
Comment (puzzleshark) [reply xir/xir.lark]: good call, not really the right word. changed to `ARBITRARY_NUM_WIRES`

File: xir/parser.py (hunk 10 above)
Comment (puzzleshark) [reply xir/parser.py]: fixed

File: xir/parser.py (hunk 12 above)
Comment (puzzleshark) [reply xir/parser.py]: changed

File: xir/parser.py (hunk 13 above)
Comment (puzzleshark) [reply xir/parser.py]: changed the naming of this constant to be more clear on the purpose

File: xir/xir.lark (hunk 16 above)
[the remark this answers is not in the document]
Comment (Mandrenkov) [reply xir/xir.lark]: Don't forget to change the name of the `wires_or_enum` rule too!

File: xir/parser.py (hunk 14 above)
Comment (Mandrenkov) [reply xir/parser.py]: Hmm, what do you mean? 🤔

File: xir/program.py [hunk 17]
```diff
@@ -226,23 +227,31 @@ def __init__(
...
+        wires: Optional[Union[Sequence[Wire], ARBITRARY_NUM_WIRES]] = None,
     ) -> None:
         if type_ not in ("gate", "out", "obs", "func"):
             raise TypeError(f"Declaration type '{type_}' is invalid.")
 
         self._name = name
         self._type = type_
         self._params = list(params or [])
-        self._wires = tuple(wires or ())
+        self._any_wires = False
```
Comment (Mandrenkov) [inline xir/program.py:238]: Where do we use this attribute?

File: xir/program.py [hunk 18]
```diff
@@ -24,6 +24,8 @@
 Param = Union[complex, str, Decimal, DecimalComplex, bool, List["Param"]]
 Params = Union[List[Param], Dict[str, Param]]
 
+ARBITRARY_NUM_WIRES = "..."
```
Comment (Mandrenkov) [inline xir/program.py:27]: We wouldn't need a constant like this if we just used `...` directly, right?

File: xir/parser.py (hunk 11 above)
Comment (Mandrenkov) [reply xir/parser.py:246]: Hmm, have you checked if the tests pass without the `if` statement? 🤔

**Review by Mandrenkov — COMMENTED** [review]
Nice!

As a heads up, we usually click the "refresh" icon beside a reviewer when the PR is ready for another round of review!

File: xir/parser.py [hunk 19]
```diff
@@ -275,69 +234,74 @@ def application_stmt(self, args):
...
+    @v_args(inline=True)
+    def gate_decl(self, name, params, wires):
         """Gate declaration. Adds declaration to program."""
-        if len(args) == 3:
-            name, params, wires = args[0], args[1][1], args[2][1]
-        else:
-            name, wires = args[0], args[1][1]
-            params = []
-
+        params = params[1] if params else []
```
Comment (thisac) [inline xir/parser.py:277]: This is so much cleaner! 🚀

File: tests/test_program.py [hunk 20]
```diff
@@ -127,8 +127,8 @@ def test_statements_no_params(self, program, name, wires):
     def test_observable_stmt(self, program, pref, wires):
         """Tests serializing an XIR program with observable statements."""
         xyz = "XYZ"
-        terms = [(xyz[i], w) for i, w in enumerate(wires)]
-        terms_str = " @ ".join(f"{t[0]}[{t[1]}]" for t in terms)
+        terms = [xir.ObservableFactor(xyz[i], None, w) for i, w in enumerate(wires)]
+        terms_str = " @ ".join(str(t) for t in terms)
```
Comment (thisac) [inline tests/test_program.py:131]: Might as well change `terms` to `factors` in the tests as well (including in several places below).
```suggestion
        factors = [xir.ObservableFactor(xyz[i], None, w) for i, w in enumerate(wires)]
        factors_str = " @ ".join(str(t) for t in factors)
```

**Review by thisac — CHANGES_REQUESTED** [review]
Nice work @puzzleshark! 🚀 The inline arguments make things much nicer, although I'm a bit hesitant to keep them if they're the cause for the performance slowdown. This is definitely something that we should look into before merging this PR.

File: xir/parser.py (hunk 11 above)
Comment (Mandrenkov) [reply xir/parser.py:246]: Oh, true! I suppose there are a couple of workarounds:
```suggestion
                    max_wire = max(max_wire, max_wire, *int_wires)
```
or
```suggestion
                    max_wire = max((max_wire, *int_wires))
```

File: xir/__init__.py [hunk 21]
```diff
@@ -17,8 +17,15 @@
...
 
-def parse_script(script: str, debug=True, **kwargs) -> Program:
-    """Parses an XIR script into a structured :class:`xir.Program`."""
+def parse_script(script: str, debug: bool = True, **kwargs) -> Program:
+    """
+    Parses an XIR script into a structured :class:`xir.Program`.
+
+    script (str): xir script as a string.
+    debug (bool): if false lark tree building will be skipped, and lark rule collisions will not be
+    given a warning.
```
[the remark this answers is not in the document]
Comment (puzzleshark) [reply xir/__init__.py:26]: I actually think we should remove the kwargs. what do you think?

File: xir/parser.py [hunk 22]
```diff
@@ -254,19 +213,19 @@ def application_stmt(self, args):
             if a == "inv":
                 inverse = not inverse
             elif a == "ctrl":
-                ctrl_wires.update(args.pop(0)[1])
+                ctrl_wires.update(args.pop(0))
```
[the remark this answers is not in the document]
Comment (puzzleshark) [reply xir/parser.py:216]: because wires used to be a tuple `("wires", (0, 1, 2))` now it's just a tuple

File: xir/parser.py [hunk 23]
```diff
@@ -275,69 +234,74 @@ def application_stmt(self, args):
...
 
-        Returns:
-            list[tuple]: each observable with corresponding wires as tuples
-        """
-        return [(args[i], args[i + 1]) for i in range(0, len(args) - 1, 2)]
+        params = params[1] if params else []
+        self._program.add_observable(name, params, wires, statements)
+
+    def obs_stmt_list(self, stmts):
+        return stmts
```
[the remark this answers is not in the document]
Comment (puzzleshark) [reply xir/parser.py:257]: the issue with that is now in your Transformer you need to have something like `def obs_def(name, params_list, wires, *obs_stmt)` which I don't think is a good idea. because this is dependent on the order of the params in the definition. if the rule was `obs_def: _OP name [params_list] : obs_stmt+ _END [wires] ";"` then it wouldn't work

File: xir/parser.py [hunk 24]
```diff
@@ -275,69 +234,74 @@ def application_stmt(self, args):
...
+
+    @v_args(inline=True)
+    def obs_stmt(self, pref, factors):
+        return ObservableStmt(simplify_math(pref), factors, use_floats=self.use_floats)
+
+    def obs_group(self, factors):
+        return factors
```
[the remark this answers is not in the document]
Comment (puzzleshark) [reply xir/parser.py:264]: see comment above

File: xir/parser.py [hunk 25]
```diff
@@ -275,69 +234,74 @@ def application_stmt(self, args):
...
-        if len(args) == 2:
-            name, params = args[0], args[1][1]
-        else:
-            name = args[0]
-            params = []
+    def wire_list(self, args):
+        return args
+
+    def ARBITRARY_NUM_WIRES(self, _):
+        return ARBITRARY_NUM_WIRES
```
[the remark this answers is not in the document]
Comment (puzzleshark) [reply xir/parser.py:292]: this could create a problem, if "..." was changed in lark file, tests would fail

[the remark this answers is not in the document]
Comment (puzzleshark) [reply xir/parser.py:292]: `ARBITRARY_NUM_WIRES` can really be anything, it just needs to be a unique constant because we do checks during the validation steps

File: xir/program.py (hunk 18 above)
Comment (puzzleshark) [reply xir/program.py:27]: see my comment above https://github.com/XanaduAI/xir/pull/6/files#r762130077

File: xir/program.py [hunk 26]
```diff
@@ -137,27 +139,54 @@ def use_floats(self) -> bool:
...
+    Args:
+        name (str): name of the factor
+        params (Params): the parameters of the factor
+        wires (Sequence[Wire]): the wires this factor acts on
+    """
+
+    def __init__(self, name: str, params: Params, wires: Sequence[Wire]) -> None:
+        self.name = name
+        self.params = list(params or [])
+        self.wires = tuple(wires or [])
```
[the remark this answers is not in the document]
Comment (puzzleshark) [reply xir/program.py:154]: good call, I wasn't thinking about the `dict` case

File: xir/program.py (hunk 17 above)
Comment (puzzleshark) [reply xir/program.py:238]: You're right, it's not needed

File: xir/xir.lark [hunk 27]
```diff
@@ -13,23 +13,29 @@ option: name ":" val
...
 obs_stmt: expr "," obs_group ";"
-obs_group: obs "[" wire "]" ("@" obs "[" wire "]")*
+obs_group: obs_factor ("@" obs_factor)*
+obs_factor: obs [params] wires
 ?obs: name
 
 // wires and parameters
-wires : "[" (wire | range_) ("," (wire | range_))* "]"
+?wires_or_enum: "[" wire_list "]" | "[" ARBITRARY_NUM_WIRES "]"
+?wires: "[" wire_list "]"
```
[the remark this answers is not in the document]
Comment (puzzleshark) [reply xir/xir.lark:37]: good idea

File: xir/__init__.py (hunk 21 above)
[the remark this answers is not in the document]
Comment (puzzleshark) [reply xir/__init__.py]: actually keeping it in for now. however in the future for optimal performance we might need to create the transformer in advance, and might need a slight work around

File: xir/parser.py (hunk 14 above)
Comment (puzzleshark) [reply xir/parser.py]: oh just that, if you want to refactor like change "..." to "enum" then it's more complicated since "..." is in multiple places vs just one.

File: xir/program.py (hunk 4 above)
Comment (Mandrenkov) [reply xir/program.py]: Yeah, we should avoid over-constraining types where possible. This results in a more flexible interface so that
```python
def avg(s: Sequence[int]) -> int:
    return sum(s) / len(s)
```
can be called using
```python
avg([0, 1, 2, 3, 4])
```
or
```python
avg((0, 1, 2, 3, 4))
```
or even
```python
avg(range(5))
```
[This page](https://docs.python.org/3/library/collections.abc.html#collections-abstract-base-classes) is a pretty handy reference.

File: xir/parser.py (hunk 14 above)
Comment (Mandrenkov) [reply xir/parser.py]: That's true, but I would still argue that, as a user of the XIR, it is more intuitive to deal with `...` objects then a string constant which is meant to represent a similar thing. Also, I wouldn't really expect the representation of the "arbitrary number of wires" object to change so there isn't much need to design our code around that use case.

File: xir/program.py (hunk 18 above)
Comment (Mandrenkov) [reply xir/program.py:27]: Okay, let's continue this discussion in https://github.com/XanaduAI/xir/pull/6#discussion_r769687832.

**Review by Mandrenkov — COMMENTED** [review]
I'm still not entirely convinced about the `enum` name and not using `...` but otherwise this looks good!

File: xir/xir.lark [hunk 28]
```diff
@@ -33,8 +33,9 @@ obs_factor: obs [params] wires
 ?obs: name
 
 // wires and parameters
-?wires_or_enum: "[" wire_list "]" | "[" ARBITRARY_NUM_WIRES "]"
+?wires_or_enum: (wires | enum)
 ?wires: "[" wire_list "]"
+?enum: "[" ARBITRARY_NUM_WIRES "]"
```
[the remark this answers is not in the document]
Comment (puzzleshark) [reply xir/xir.lark]: good call I forgot to change it in the lark file. changed to arbitrary

**Review by Mandrenkov — APPROVED** [review]
Looks good to me assuming we've addressed the root cause of the benchmark regression!

File: xir/parser.py (hunk 22 above)
[the remark this answers is not in the document]
Comment (thisac) [reply xir/parser.py:207]: Oh, right! That's much nicer! 💯

File: xir/parser.py [hunk 29]
```diff
@@ -275,69 +225,98 @@ def application_stmt(self, args):
...
-
-        Returns:
-            ObservableStmt: object containing statement data
+    @v_args(inline=True)
+    def obs_def(self, name, params, wires, statements):
         """
-        pref = simplify_math(args[0])
-        terms = args[1]
-        return ObservableStmt(pref, terms, use_floats=self.use_floats)
+        Creates an observable definition from text of the form:
```
Comment (thisac) [inline xir/parser.py:231]: (minor) the first sentence should be a single line (ending with a `.`) on the same line as the `"""`. This can then be followed by a more detailed description. I realize that I didn't follow this convention fully before either.
```python
"""Creates an observable definition and adds it to the program.

Creates an observable definition from text of the form:
...
"""
```
No a big deal here, since these docstring currently aren't rendered in the readthedocs.

File: xir/parser.py [hunk 30]
```diff
@@ -275,69 +225,98 @@ def application_stmt(self, args):
...
 
-    def obs_group(self, args):
-        """Group of observables used to define an observable statement.
+        .. code-block:: text
+
+            obs my_obs(params)[0, 1]:
+                1, obs_1[0];
+                0.5, obs_2[1];
+            end;`
```
Comment (thisac) [inline xir/parser.py:238]:
```suggestion
            end;
```

File: xir/parser.py [hunk 31]
```diff
@@ -275,69 +225,98 @@ def application_stmt(self, args):
...
+
+        Args:
+            name: observable name.
+            params: observable params.
+            wires: observable wires.
+            statements: list of statements.
```
Comment (thisac) [inline xir/parser.py:244]:
```suggestion
            name: observable name
            params: observable params
            wires: observable wires
            statements: list of statements
```

**Review by thisac — CHANGES_REQUESTED** [review]
Looking really good! Just three things that I think should be done before merging:
* Move the parser creation outside of the `parse_script` function to avoid having to create the parser each time a user wants to parse a script.
* Add tests for the new `...` syntax, both unit and integration (I couldn't see any).
* Update the changelog.

I'll approve as soon as these things are addressed. Otherwise, looks great! 💯

File: xir/program.py (hunk 18 above)
Comment (thisac) [reply xir/program.py:27]: I'm not sure why I cannot reply to the other discussion thread (there simply is no reply-box!), so I'll just write here instead. 🤷 

> That's true, but I would still argue that, as a user of the XIR, it is more intuitive to deal with `...` objects then a string constant which is meant to represent a similar thing. Also, I wouldn't really expect the representation of the "arbitrary number of wires" object to change so there isn't much need to design our code around that use case.

I'm leaning towards agreeing with @Mandrenkov here, since we'll likely not update this more often than any of any other syntax. This would be closer to how we handle e.g., the inverse and ctrl modifiers.

File: xir/__init__.py [hunk 32]
```diff
@@ -1,13 +1,59 @@
...
-    tree = parser.parse(script)
-    return Transformer(**kwargs).transform(tree)
+    parser = lark.Lark(
+        grammar=_read_lark_file(),
+        maybe_placeholders=True,
+        start="program",
+        parser="lalr",
+        debug=False,
+        transformer=Transformer(**kwargs),
+    )
```
[the remark this answers is not in the document]
Comment (puzzleshark) [reply xir/__init__.py]: fixed

File: xir/__init__.py [hunk 33]
```diff
@@ -1,13 +1,59 @@
...
+def parse_script(script: str, debug: bool = False, **kwargs) -> Program:
+    """
+    Parses an XIR script into a structured :class:`xir.Program`.
+
+    Args:
+        script (str): xir script as a string.
+        debug (bool): if false lark tree building will be skipped, and lark rule collisions
+        will not be given a warning.
+        kwargs: options to be passed to the transformer.
+    """
```
[the remark this answers is not in the document]
Comment (puzzleshark) [reply xir/__init__.py]: fixed

File: xir/parser.py (hunk 29 above)
Comment (puzzleshark) [reply xir/parser.py]: gotcha, didn't know this. thanks for the info!

File: xir/__init__.py [hunk 34]
```diff
@@ -1,13 +1,88 @@
...
+
+    Args:
+        debug (bool): if false lark tree building will be skipped, and lark rule collisions
+        will not be given a warning.
+        kwargs: options to be passed to the transformer.
+    Returns:
+        a parsing function.
+    """
+
+    def _inner_script_parser(script):
```
Comment (thisac) [inline xir/__init__.py:43]: Unfortunately, I think that the caching doesn't work here as expected. We'd like to separate the parser creation from the parsing itself so that the parser creation can be cached while still being able to pass different scripts to be passed. Something like the following should work, I believe:
```python
@lru_cache()
def _get_parser(debug: bool = False, **kwargs):
    """Create parser from options.

    Args:
        debug (bool): if false lark tree building will be skipped, and lark rule collisions
            will not be given a warning

    Returns:
        Callable: a parsing function
    """
    trans = Transformer(**kwargs) if not debug else None

    parser = lark.Lark(
        grammar=_read_lark_file(),
        maybe_placeholders=True,
        start="program",
        parser="lalr",
        debug=debug,
        transformer=trans,
    )

    def _inner_script_parser(script) -> Program:
        """Parse a script.

        Args:
            script (str): xir script as a string

        Returns:
            xir.Program: program representation of the script
        """
        if debug:
            tree = parser.parse(script)
            return Transformer(**kwargs).transform(tree)
        return parser.parse(script)

    return _inner_script_parser
```

**Review by thisac — COMMENTED** [review]
Thanks @puzzleshark! I think that the caching might not be working as expected, so we still have the slowdown of having to create the parser each time that `parse_script` is called with different scripts. I added a suggestion for how to solve this. 🙂

File: xir/__init__.py (hunk 34 above)
Comment (thisac) [reply xir/__init__.py:43]: @puzzleshark I noticed an issue with my suggestion above. When using `debug=False` it seems like the parser keeps the previously parsed script in memory, and simply appends the new script to it. This isn't the case when `debug=True` though. E.g., 

```python
import xir

xir_script1 = "Rgate(a) | [0];"
xir_script2 = "BSgate(b, c) | [0, 1];"

prog1 = xir.parse_script(xir_script1)
prog2 = xir.parse_script(xir_script2)

print(prog1.serialize())
```

outputs

```
Rgate(a) | [0];
BSgate(b, c) | [0, 1];
```

which I find slightly puzzling, since both `prog1` and `prog2` turn out to be equal.

Comment (puzzleshark) [reply xir/__init__.py:60]: hmm good point, I think the issue is being caused by creating a `Program()` instance in the constructor of the Transformer class. This requires a more careful design because as it is it's not so easy to remove

Comment (puzzleshark) [reply xir/__init__.py:60]: I think maybe we should consider, for now for `debug=False` create a new parser for each script. and for `debug=False` use the method you suggested, this will make it on par with `main`. Maybe in the future we can consider changing how a `Program` is built.

Comment (thisac) [reply xir/__init__.py:64]: > I think maybe we should consider, for now for `debug=False` create a new parser for each script. and for `debug=False` use the method you suggested

I guess you mean `debug=True` for the latter? I'd be OK with that, as long as we remember to look into this at a later stage (due to the increased time having to creating the parser each time a script is parsed).

Comment (puzzleshark) [reply xir/__init__.py:54]: resolved as discussed

File: xir/__init__.py [hunk 35]
```diff
@@ -1,13 +1,82 @@
...
+def _get_parser(debug: bool = False, **kwargs):
+    """Create parser from options.
+
+    Args:
+        debug (bool): if false lark tree building will be skipped, and lark rule collisions
+        will not be given a warning
+        kwargs: options to be passed to the transformer
+    Returns:
+        a parsing function
+    """
```
Comment (thisac) [inline xir/__init__.py:40]:
```suggestion
        kwargs: options to be passed to the transformer

    Returns:
        Callable: a parsing function
    """
```

**Review by thisac — APPROVED** [review]
Fantastic job @puzzleshark! I'm happy to have this merged in. 🚀 💯
